source/vistas/core/graphics/feature/factory.py

```diff
@@ -0,0 +1,234 @@
+import os
+from functools import partial
+
+import numpy
+import pyproj
+import shapely.geometry as shp
+from shapely.ops import transform
+from triangle import triangulate
+
+from vistas.core.color import RGBColor
+from vistas.core.gis.elevation import ElevationService, TILE_SIZE, meters_per_px
+from vistas.core.graphics.factory import MapMeshFactory, MeshFactoryWorker
+from vistas.core.graphics.feature.geometry import FeatureGeometry
+from vistas.core.graphics.feature.shader import FeatureShaderProgram
+from vistas.core.graphics.mesh import Mesh
+from vistas.core.plugins.data import FeatureDataPlugin
+
+
+class FeatureFactoryWorker(MeshFactoryWorker):
+
+    task_name = "Building Features"
+
+    def work(self):
+        verts, indices, normals, colors = [None] * 4
+        if self.factory.needs_vertices and not self.task.should_stop:
+            verts, indices, normals = self.factory.generate_meshes(self.task)
+            self.factory.needs_vertices = False
+
+        if self.factory.needs_color and not self.task.should_stop:
+            colors = self.factory.generate_colors(self.task)
+            self.factory.needs_color = False
+
+        self.sync_with_main(
+            self.factory.update_features, kwargs=dict(vertices=verts, indices=indices, normals=normals, colors=colors),
+            block=True
+        )
+
+
+class FeatureFactory(MapMeshFactory):
+    """ A MapMeshFactory for handling polygon rendering. """
+
+    worker_class = FeatureFactoryWorker
+
+    def __init__(self, extent, data_src: FeatureDataPlugin, shader=None, plugin=None, initial_zoom=10):
+        if shader is None:
+            shader = FeatureShaderProgram()
+        super().__init__(extent, shader, plugin, initial_zoom)
+
+        self._color_func = None
+        self._render_thread = None
+
+        assert isinstance(data_src, FeatureDataPlugin)
```

Comment:
While technically, this is an [appropriate use](https://wiki.python.org/moin/UsingAssertionsEffectively) of `assert`, I personally prefer _not_ to use it validate parameters for a few reasons: 1) it might get switched off (in production, asserts can be disabled); 2) it's semantically unclear (all asserts are the same, and while a failed assert may provide useful data to a programmer, it's very vague, basically "this is unexpected"); 3) used outside of testing, assert feels temporary, like a first draft.

Instead, I prefer to use `TypeError` when the parameter type is incorrect, or `ValueError` if the parameter is of the correct type, but is wrong (e.g., if your function only takes prime numbers and receives a non-prime number). This is more specific, and allows for an added message to explain further if necessary: `raise TypeError('This function requires a `Foo` because reasons')`.

An exception might be if the assert should not cause a failure, _except_ when in debug; i.e., the code will work just fine in production, but you want to chastise a developer because it *might* be problematic.

\*shrug\* maybe I'm alone here; in which case go ahead and ignore this. Assert just doesn't feel like the right tool to me.

Reply:
I'm good with using `TypeError`, and it would be bad if it were such that a non-`FeatureDataPlugin` was passed to the `FeatureFactory` class. I'll swap it out for a `TypeError`.